src/permissions.js

```diff
@@ -0,0 +1,60 @@
+/**
```

Comment:
When I look at this I keep feeling like there must be a way to make it more concise, or that it should early-return, etc. but I think the fact that it is very specific about the warnings and gives a full report of problems is pretty nice to the caller. It feels repetitive but having one super generic function and 4 callers that will never be expanded doesn't seem like it adds any clarity.

Reply:
Anyway, nothing to resolve here really just puzzled over it for a little bit, but I think it's good as-is. Nice work!